### PR TITLE
Fix benchmarks in library/core with black_box

### LINKED ISSUE #107590: Wrong benchmarks in library/core/benches/char/methods.rs and possibly other places
As I learned from https://gendignoux.com/blog/2022/01/31/rust-benchmarks.html#application-to-real-benchmarks file `library/core/benches/char/methods.rs` and possibly other files in this repo contain wrong benchmarks.

As author of the article points, call `c.to_digit(2)` optimizes ~to None~ and thus doesn't get benchmarked.

So, please, find all such places and fix them. The faulty example still present in latest master

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #107590

### CODE REVIEW

File: library/core/benches/char/methods.rs [hunk 1]
```diff
@@ -1,26 +1,26 @@
-use test::Bencher;
+use test::{black_box, Bencher};
 
 const CHARS: [char; 9] = ['0', 'x', '2', '5', 'A', 'f', '7', '8', '9'];
 const RADIX: [u32; 5] = [2, 8, 10, 16, 32];
 
 #[bench]
 fn bench_to_digit_radix_2(b: &mut Bencher) {
-    b.iter(|| CHARS.iter().cycle().take(10_000).map(|c| c.to_digit(2)).min())
+    b.iter(|| black_box(CHARS.iter().cycle().take(10_000).map(|c| black_box(c).to_digit(2)).min()))
```
Comment (gendx) [inline library/core/benches/char/methods.rs:8]: There is no need to put a `black_box` on the output, as it is automatically added by the `Bencher` interface: https://github.com/rust-lang/rust/blob/1.67.0/library/test/src/bench.rs#L117. The important part (for the output) is that the lambda function passed as argument to `.iter()` returns a non-`()` result (which is already the case here).

The `black_box` is definitely relevant for the input though :)

**Comment by thomcc** [comment]
I agree with @gendx's comment. Can you move black_box to be on the inputs only? (Then comment with at-rustbot ready to put it back in my queue).

@rustbot author

**Comment by thomcc** [comment]
Looks good.

@bors r+ rollup

**Pull request merged**